Support arbitrary blob signing (#283)
notaryproject/notaryproject: specs/trust-store-trust-policy.md

Any additional sub directories under a named store and certificates in it are ignored. **NOTE**: Implementation SHOULD warn if it finds sub directories with certificates under a named store, to help diagnose misconfigured store.

## Trust Policy

Users who consume signed artifacts from OCI registries, or signed arbitrary blobs stored on file system, use trust policies to specify trusted identities that signed the artifacts, and level of signature verification to enforce. The trust policy is a JSON document.

**NOTE**: Support for verifying signed arbitrary blob has been added in policy version 1.1. While the policy version 1.1 can be used for verifying both signed OCI artifacts as well as signed arbitrary blobs, policy version 1.0 can only be used for verifying signed OCI artifacts.

### Trust Policy Schema

#### Version 1.1

- **`version`**(*string*): This REQUIRED property is the version of the trust policy. This MUST be either `1.1` or `1.0`.
- **`trustPolicies`**(*string-array of objects map*): This REQUIRED property represents a collection of trust policies.
  - **`name`**(*string*): This REQUIRED property represents the name of the trust policy.

>   - **`scopes`**(*array of strings*): This REQUIRED property acts as a policy selector and determines which trust policy is applicable for verifying a given artifact. Values MUST be prefixed with either `oci:` or `blob:` to indicate what artifact type the scope must be applied to.

Reviewer: Can scopes value contain random prefix like `abc:sometext`? Is this a valid trust policy statement ?

Author: This comment is irrelevant with the new proposal of separating policy files.

            "name": "wabbit-networks-images",   // Name of the policy.
            "scopes": [ "oci:*" ],              // The registry artifacts to which the policy applies.
            "signatureVerification": {          // The level of verification - strict, permissive, audit, skip.
              "level" : "audit"
            },
            "trustStores": ["ca:acme-rockets"], // The trust stores that contains the X.509 trusted roots.
            "trustedIdentities": [              // Identities that are trusted to sign the artifact.
              "x509.subject: C=US, ST=WA, L=Seattle, O=acme-rockets.io, OU=Finance, CN=SecureBuilder"
            ]
        }
    ]
}
```

2.Trust policy for a scenario where ACME Rockets consumes some OCI artifacts signed by Wabbit Networks and some signed by ACME Rockets.

```jsonc
{
    "version": "1.1",
    "trustPolicies": [
        {
            // Policy for set of artifacts signed by Wabbit Networks
            // that are pulled from ACME Rockets repository
            "name": "wabbit-networks-images",
            "scopes": [
              "oci:registry.acme-rockets.io/software/net-monitor",

>               "oci:registry.acme-rockets.io/software/net-logger"

Reviewer 2: Can we mix `oci:` and `blob:`?

Author: Nope. `Scopes` constraints already clarifies this.

The Trust store currently supports three kinds of identities, additional identities may be supported in future :

- **Certificates**: The `x509/ca` trust store contains named stores that contain Certificate Authority (CA) root certificates.
- **SigningAuthority Certificate**: The `x509/signingAuthority` trust store contains named stores that contain Signing Authority's root certificates.
- **Timestamping Certificates**: The `x509/tsa` trust store contains named stores with Time Stamping Authority (TSA) root certificates.

Any additional sub directories under a named store and certificates in it are ignored. **NOTE**: Implementation SHOULD warn if it finds sub directories with certificates under a named store, to help diagnose misconfigured store.

## Trust Policy

Users who consume signed artifacts from OCI registries, or signed arbitrary blobs stored on file system, use trust policies to specify trusted identities that signed the artifacts, and level of signature verification to enforce. Trust policy files are JSON documents.

> **NOTE**: Users are required to setup separate trust policy files for verifying different types of signed artifacts. Notation expects OCI trust policy file to be placed at `{NOTATION_CONFIG}/trustpolicy.oci.json`. For backward compatibility reasons, Notation accepts `{NOTATION_CONFIG}/trustpolicy.json` as OCI trust policy file as well. For arbitrary blobs, Notation expects trust policy file to be placed at `{NOTATION_CONFIG}/trustpolicy.blob.json`.

Author: @priteshbandi @shizhMSFT 

The original proposal has `trustpolicy.oci-image.json` for OCI trust policy. But, I'm changing it to `trustpolicy.oci.json` to keep it bit more generic so that this trust policy can handle all types of artifacts under "OCI" umbrella rather than just images. LMK your thoughts.

> **NOTE**: Users are required to setup separate trust policy files for verifying different types of signed artifacts. Notation expects OCI trust policy file to be placed at `{NOTATION_CONFIG}/trustpolicy.oci.json`. For backward compatibility reasons, Notation accepts `{NOTATION_CONFIG}/trustpolicy.json` as OCI trust policy file as well. For arbitrary blobs, Notation expects trust policy file to be placed at `{NOTATION_CONFIG}/trustpolicy.blob.json`.

Author: Original `{NOTATION_CONFIG}/trustpolicy.json` is still allowed. Are you suggesting to remove `{NOTATION_CONFIG}/trustpolicy.oci.json`?

> **NOTE**: Users are required to setup separate trust policy files for verifying different types of signed artifacts. Notation expects OCI trust policy file to be placed at `{NOTATION_CONFIG}/trustpolicy.oci.json`. For backward compatibility reasons, Notation accepts `{NOTATION_CONFIG}/trustpolicy.json` as OCI trust policy file as well. For arbitrary blobs, Notation expects trust policy file to be placed at `{NOTATION_CONFIG}/trustpolicy.blob.json`.

Author: It can be confusing to some users who do both OCI and blob verification as to why trustpolicy.json is for OCI. They can use `{NOTATION_CONFIG}/trustpolicy.oci.json` and `{NOTATION_CONFIG}/trustpolicy.blob.json` for disambiguation.

> **NOTE**: Users are required to setup separate trust policy files for verifying different types of signed artifacts. Notation expects OCI trust policy file to be placed at `{NOTATION_CONFIG}/trustpolicy.oci.json`. For backward compatibility reasons, Notation accepts `{NOTATION_CONFIG}/trustpolicy.json` as OCI trust policy file as well. For arbitrary blobs, Notation expects trust policy file to be placed at `{NOTATION_CONFIG}/trustpolicy.blob.json`.

Reviewer 2: nit: you might want to use

```
> [!NOTE]
> here is a note
```
> [!NOTE]
> here is a note

Instead of `**NOTE**:`.

### Trust Policy Properties

#### OCI Trust Policy

##### Version 1.0

> - **`version`**(*string*): This REQUIRED property is the version of the trust policy. This MUST be `1.0` at the moment.

Reviewer: Should we have common version for trust policy irrespective of blob or OCI ?

Author: Discussed offline. blob and oci policies may evolve differently and have different version numbers.

#### Blob Trust Policy

##### Version 1.0

- **`version`**(*string*): This REQUIRED property is the version of the trust policy. This MUST be `1.0` at the moment.
- **`trustPolicies`**(*string-array of objects map*): This REQUIRED property represents a collection of trust policies.
  - **`name`**(*string*): This REQUIRED property represents the name of the trust policy. This will be used to select the policy to verify a signed blob.
  - **`signatureVerification`**(*object*): This REQUIRED property dictates how signature verification is performed.
  An *object* that specifies a predefined verification level, with an option to override the Notary Project trust policy defined verification level if user wants to specify a [custom verification level](#custom-verification-level).
    - **`level`**(*string*): A REQUIRED property that specifies the verification level, supported values are `strict`, `permissive`, `audit` and `skip`. Detailed explanation of each level is present [here](#signatureverification-details).
    - **`override`**(*map of string-string*): This OPTIONAL map is used to specify a [custom verification level](#custom-verification-level).

>   - **`wildcardPolicy`**(*boolean*): This OPTIONAL property flags the policy as the wildcard trust policy if set to `true`. This policy will be used for verification if the user does not select any policy by its name during verification.

Reviewer: nit: should key be `isWildCardPolicy` ?

Author: Changed it to `globalPolicy`. I think this level of detail is sufficient for the readers.

The Trust store currently supports three kinds of identities, additional identities may be supported in future :

- **Certificates**: The `x509/ca` trust store contains named stores that contain Certificate Authority (CA) root certificates.
- **SigningAuthority Certificate**: The `x509/signingAuthority` trust store contains named stores that contain Signing Authority's root certificates.
- **Timestamping Certificates**: The `x509/tsa` trust store contains named stores with Time Stamping Authority (TSA) root certificates.

Any additional sub directories under a named store and certificates in it are ignored. **NOTE**: Implementation SHOULD warn if it finds sub directories with certificates under a named store, to help diagnose misconfigured store.

## Trust Policy

> Users who consume signed artifacts from OCI registries or signed arbitrary blobs with detached signatures use trust policies to specify trusted identities that signed the artifacts, and level of signature verification to enforce. The trust policy is a JSON document.

Reviewer 3: ```suggestion
Users who consume signed artifacts from OCI registries, or signed arbitrary blobs stored on file system, use trust policies to specify trusted identities that signed the artifacts, and level of signature verification to enforce. The trust policy is a JSON document.
```

**NOTE**: Support for verifying signed arbitrary blob has been added in policy version 1.1. While the policy version 1.1 can be used for verifying both signed OCI artifacts as well as signed arbitrary blobs, policy version 1.0 can only be used for verifying signed OCI artifacts.

### Trust Policy Schema

#### Version 1.1

- **`version`**(*string*): This REQUIRED property is the version of the trust policy. This MUST be either `1.1` or `1.0`.
- **`trustPolicies`**(*string-array of objects map*): This REQUIRED property represents a collection of trust policies.
  - **`name`**(*string*): This REQUIRED property represents the name of the trust policy.

>   - **`scopes`**(*array of strings*): This REQUIRED property acts as a policy selector and determines which trust policy is applicable for verifying a given artifact. Values MUST be prefixed with either `oci:` or `blob:` to indicate what artifact type the scope must be applied to.

Reviewer 3: Can the scopes collection contain both oci and blob prefix values?

Author: Do not support mixing different scopes..

Reviewer 3: Updated to be more explicit

```suggestion
  - **`scopes`**(*array of strings*): This REQUIRED property acts as a policy selector and determines which trust policy is applicable for verifying a given artifact. Values MUST be prefixed with either `oci:` or `blob:` to indicate what artifact type the scope must be applied to. All values in the array MUST be prefixed with either `oci:`, or all values MUST be prefixed with `blob:`, prefixes cannot be mixed in a the same policy.
```

Reviewer 2: May I know the underlying reason why we don't support mixing different scopes?

Author: As we are separating OCI and Blob policies, this comments becomes irrelevant

  - **`trustStores`**(*array of string*): This REQUIRED property specifies a set of one or more named trust stores, each of which contain the trusted roots against which signatures are verified. Each named trust store uses the format `{trust-store-type}:{named-store}`. Currently supported values for `trust-store-type` are `ca`, `signingAuthority` and `tsa`. For publicly trusted TSA, `tsa:publicly-trusted-tsa` is the default value, and implied without explicitly specifying it. If a custom TSA is used the format `ca:acme-rockets,tsa:acme-tsa` is supported to specify it.
  - **`trustedIdentities`**(*array of strings*): This REQUIRED property specifies a set of identities that the user trusts. For X.509 PKI, it supports list of elements/attributes of the signing certificate's subject. For more information, see [identities constraints](#trusted-identities-constraints) section. A value `*` is supported if user trusts any identity (signing certificate) issued by the CA(s) in `trustStore`.

Note: Version 1.0 does not support verifying signed arbitrary blobs. See [Version 1.1](#version-1.1) for that.

> ### Trust Policy Examples

Reviewer 3: I'd also add a section "Migrating from policy version 1.0 to 1.1" before this section.

Reviewer 3: Add two sections under "Trust Policy Examples", "Trust Policy for OCI artifacts" and "Trust Policy for Blobs".

            // It logs expiry and skips revocation check for a specific OCI artifact.
            "name": "use-expired-image",
            "scopes": [ "oci:registry.acme-rockets.io/software/legacy/metrics" ],
            "signatureVerification": {
              "level" : "strict",
              "override" : {
                "expiry" : "log",
                "revocation" : "skip"
              }
            },
            "trustStores": ["ca:acme-rockets"],
            "trustedIdentities": ["*"]
        },
        {
            // Policy for all other OCI artifacts signed by ACME Rockets
            // from any registry location. The policy also specified multiple trust stores.
            "name": "global-policy-for-all-other-images",
            "scopes": [ "oci:*" ],
            "signatureVerification": {
              "level" : "audit"
            },
            "trustStores": ["ca:acme-rockets", "ca:acme-rockets-ca2"],
            "trustedIdentities": [
              "x509.subject: C=US, ST=WA, L=Seattle, O=acme-rockets.io, OU=Finance, CN=SecureBuilder"
            ]
        }
    ]
}
```

> 3.Trust policy in version 1.0 for a scenario where ACME Rockets consumes some OCI artifacts signed by Wabbit Networks and some signed by ACME Rockets.

Reviewer 3: I'd change 1.0 policy examples as a sub heading so that it's easy to find, and retain all the previous examples in 1.0

            "trustedIdentities": ["*"]
        },
        {
            // Policy for all other OCI artifacts signed by ACME Rockets
            // from any registry location. The policy also specified multiple trust stores.
            "name": "global-policy-for-all-other-images",
            "registryScopes": [ "*" ],
            "signatureVerification": {
              "level" : "audit"
            },
            "trustStores": ["ca:acme-rockets", "ca:acme-rockets-ca2"],
            "trustedIdentities": [
              "x509.subject: C=US, ST=WA, L=Seattle, O=acme-rockets.io, OU=Finance, CN=SecureBuilder"
            ]
        }
    ]
}
```

#### Trust Policy for Blobs

1. Trust policy for the scenario where ACME Rockets consumes arbitrary blobs signed by Wabbit Networks and some signed by ACME Rockets.

```jsonc
{
    "version": "1.0",
    "trustPolicies": [
        {
            // Policy for set of blobs signed by Wabbit Networks
            "name": "wabbit-networks-blobs",
            "signatureVerification": {
              "level" : "strict"
            },

>             "trustStores": ["wabbit-networks"],

Author: Good catch

- `strict` : Signature verification is performed at `strict` level, which enforces all validations. If any of these validations fail, the signature verification fails. This is the recommended level in environments where a signature verification failure does not have high impact to other concerns (like application availability). It is recommended that build and development environments where images are initially created, or for high assurance at deploy time use `strict` level.
- `permissive` : The `permissive` level enforces most validations, but will only logs failures for revocation and expiry. The `permissive` level is recommended to be used if signature verification is done at deploy time or runtime, and the user only needs integrity and authenticity guarantees.
- `audit` : The `audit` level only enforces signature integrity if a signature is present. Failure of all other validations are only logged.

> - `skip` : The `skip` level does not perform any signature verification. This is useful when an application uses multiple artifacts, and has a mix of signed and unsigned artifacts. Note that `skip` cannot be used with a global a.k.a wildcard policy .

Reviewer: we dont use `wildcard policy`  term "wildcard policy" for registry scopes; instead, we refer to it as the global policy. Should we consider renaming "wildcard policy" to "global policy"?

Author: Changed it to Global Policy

- `strict` : Signature verification is performed at `strict` level, which enforces all validations. If any of these validations fail, the signature verification fails. This is the recommended level in environments where a signature verification failure does not have high impact to other concerns (like application availability). It is recommended that build and development environments where images are initially created, or for high assurance at deploy time use `strict` level.
- `permissive` : The `permissive` level enforces most validations, but will only logs failures for revocation and expiry. The `permissive` level is recommended to be used if signature verification is done at deploy time or runtime, and the user only needs integrity and authenticity guarantees.
- `audit` : The `audit` level only enforces signature integrity if a signature is present. Failure of all other validations are only logged.

> - `skip` : The `skip` level does not perform any signature verification. This is useful when an application uses multiple artifacts, and has a mix of signed and unsigned artifacts. Note that `skip` cannot be used with a global scope (`*`).

Reviewer 3: Suggestion - Note that `skip` cannot be used with a global scope, i.e `registryScopes` set to `*` in a Trust Policy for OCI artifacts, or if `globalPolicy` is set to `true` in a Trust Policy for Blobs.

- For a given OCI artifact, if there is no applicable trust policy then implementations of the [Notary Project verification specification](./signing-and-verification-workflow.md) MUST consider the artifact as untrusted and fail signature verification.
- The scope MUST NOT support reference expansion i.e. URIs must be fully qualified.
  E.g. the scope should be `docker.io/library/registry` rather than `registry`.
- Evaluation order of trust policies:
  1. *Exact match*: If there exists a trust policy whose scope contains the artifact's repository URI then the aforementioned policy MUST be used for signature evaluation.
     Otherwise, continue to the next step.
  1. *Global*: If there exists a trust policy with global/wildcard scope (`*`) then use that policy for signature evaluation.
     Otherwise, fail the signature verification.

#### Selecting a trust policy to verify a signed blob

- The signature verifier must select the appropriate trust policy for blob signature verification using policy names.
- For a given policy name selected by the verifier, there MUST be only one trust policy with that exact name.

> - For a given policy name selected by the verifier, if there is no matching trust policy with that name, then implementations of the [Notary Project verification specification](./signing-and-verification-workflow.md) MUST consider the blob as untrusted and fail signature verification.

Author: If there is no matching policy found, then we should fail the verification saying policy is not found. If user does not select a policy at all, then we will use the wildcard policy if present.

> - Evaluation order of trust policies:

Reviewer: There is no evaluation order in case of blob policy.
* If user specifies the policy name use that policy.
* If user doesn't specifies the policy name use wildcard/global policy.

### Steps

1. **Identify applicable trust policy**
   1. For OCI artifacts, refer to `{NOTATION_CONFIG}/trustpolicy.oci.json` or `{NOTATION_CONFIG}/trustpolicy.json` file and use [the artifact URI as the policy selector](#selecting-a-trust-policy-to-verify-a-signed-oci-artifact) and select the policy with matching scope from `registryScopes`.
   2. For Blob artifacts, refer to `{NOTATION_CONFIG}/trustpolicy.blob.json` and use [the policy name provided by the verifier](#selecting-a-trust-policy-to-verify-a-signed-blob) as the policy selector and select a policy.

>    3. If an applicable trust policy cannot be found, fail signature verification.

Reviewer: ```suggestion
   1. For OCI artifacts, refer to `{NOTATION_CONFIG}/trustpolicy.oci.json` or `{NOTATION_CONFIG}/trustpolicy.json` file and use [the artifact URI as the policy selector](#selecting-a-trust-policy-to-verify-a-signed-oci-artifact) and select the policy with matching scope from `registryScopes`.
   1. For Blob artifacts, refer to `{NOTATION_CONFIG}/trustpolicy.blob.json` and use [the policy name provided by the verifier](#selecting-a-trust-policy-to-verify-a-signed-blob) as the policy selector and select a policy.
   1. If an applicable trust policy cannot be found, fail signature verification.
```

        1. For Blob artifacts, validate that signature envelope can be parsed successfully based on the signature envelope type specified in the detached signature file extension. If no file extension is available, parse the envelope type by trying each of the [supported signature envelopes](./signature-specification.md#supported-signature-envelopes). Fail integrity check if unsuccessful.
    1. Validate that the content type indicated by the `content type` signed attribute in the signature envelope is supported.
    1. Get the signing certificate from the parsed [signature envelope](https://github.com/notaryproject/notaryproject/blob/7b7d283038/signature-specification.md#signature-envelope).
    1. Determine the signing algorithm(hash+encryption) from the signing certificate and validate that the signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements)
    1. Using the public key of the signing certificate and signing algorithm identified in the previous step, validate the integrity of the signature envelope.
    1. Verify signature `payload`

>         1. Verify the signature envelope's `payload` matches the source [`payload`](./signature-specification.md#payload) that is getting verified. Make sure the artifact's digest, media type and size match the ones present in the signature envelope.

Reviewer: This is confusing, we are saying signature's payload matches source payload but source payload link points to signature payload

1. **Proceed based on signature verification level**
   1. If `signatureVerification` level is set to `skip` in the trust policy, return success.
   1. For all other `signatureVerification` levels, `strict`, `permissive` and `audit`, perform each of the validation defined in the next sections - `integrity`, `authenticity`, `trusted timestamp`, `expiry` and `revocation`.
   1. The `signatureVerification` level defines if each validation is `enforced` or `logged`
        1. `enforced` - validation failures are treated as critical, causes the overall signature verification to fail and exit. Subsequent validations are not processed.
        1. `logged` - validation failure is logged and the next validation step is processed.
   1. A signature verification is considered successful when all validation steps are completed without critical failure.
1. **Validate Integrity.**
    1. Validate signature envelope
        1. For OCI artifacts, validate that signature envelope can be parsed successfully based on the signature envelope type specified in the `blobs[0].mediaType` attribute of the signature artifact manifest.

>         1. For Blob artifacts, validate that signature envelope can be parsed successfully based on the signature envelope type specified in the detached signature file extension. If no file extension is available, guess the envelope type by trying each of the [supported signature envelopes](./signature-specification.md#supported-signature-envelopes). Fail integrity check if unsuccessful. 

Author: Can you clarify why do you think speculate is the right verb here? Notation is not speculating the format. It is guessing and then verifying.

    1. Validate that the content type indicated by the `content type` signed attribute in the signature envelope is supported.
    1. Get the signing certificate from the parsed [signature envelope](https://github.com/notaryproject/notaryproject/blob/7b7d283038/signature-specification.md#signature-envelope).
    1. Determine the signing algorithm(hash+encryption) from the signing certificate and validate that the signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements)
    1. Using the public key of the signing certificate and signing algorithm identified in the previous step, validate the integrity of the signature envelope.
    1. Verify signature `payload`
        1. Verify the signature envelope's `payload` matches the source [`payload`](./signature-specification.md#payload) that is getting verified. Make sure the artifact's digest, media type and size match the ones present in the signature envelope.

>         1. Additionally for Blob artifacts, calculate the digest of the blob using the digest algorithm specified at `targetArtifact.payload.digest` and make sure the digests match.

Reviewer: We shouldn't parse payload until we have established authenticity of signature.

1. **Validate Revocation Status:**
    1. Validate signing identity(certificate and certificate chain) revocation status using [certificate revocation evaluation](#certificate-revocation-evaluation) section as per `signingIdentityRevocation` setting in trust-policy.
1. Perform extended validation using the applicable (if any) plugin.
1. Verify signature `payload`

>     1. Verify the signature envelope's `payload` matches the source [`payload`](./signature-specification.md#payload) that is getting verified.

Reviewer 3: I think we should move this to step 3 "Validate Integrity" and add details about how the payload is constructed for each case, and which attributes are matched.

    1. For Blob artifacts, calculate the digest of the blob using the digest algorithm specified at `targetArtifact.payload.digest` and make sure the digests match.

> 1. If present, verify user metadata/custom annotations match the signature attributes.

Reviewer 3: Not sure what this refers to, can you give an example?

Author: Add more clarification
